infrastructure/src/test/java/org/corfudb/infrastructure/redundancy/RedundancyCalculatorTest.java

```diff
@@ -0,0 +1,341 @@
+        LayoutSegment segment2 = new LayoutSegment(CHAIN_REPLICATION, 6L, 9L,
+                Collections.singletonList(stripe2));
+
+        // Not required, since A is present
+        Layout layout = createTestLayout(Arrays.asList(segment1, segment2));
+        assertThat(RedundancyCalculator.canRestoreRedundancy(layout, "A"))
+                .isFalse();
+        // Not required, since the layout consists of only one segment
+        layout = createTestLayout(Collections.singletonList(segment1));
+        assertThat(RedundancyCalculator.canRestoreRedundancy(layout, "A"))
+                .isFalse();
+        // Required, since node B is not present at the beginning
+        layout = createTestLayout(Arrays.asList(segment0, segment1, segment2));
+        assertThat(RedundancyCalculator.canRestoreRedundancy(layout, "B"))
+                .isTrue();
+        // Required, since localhost is not present in the second last segment
+        layout = createTestLayout(Arrays.asList(segment0, segment2));
+        assertThat(RedundancyCalculator.canRestoreRedundancy(layout, "localhost"))
+                .isTrue();
+
+
+    }
+
+    @Test
+    public void testCreateStateListComplex() {
+
+        LayoutStripe stripe1 = new LayoutStripe(Arrays.asList("localhost", "A", "B"));
+        LayoutStripe stripe2 = new LayoutStripe(Collections.singletonList("C"));
+        LayoutStripe stripe3 = new LayoutStripe(Collections.singletonList("D"));
+        LayoutSegment segment1 = new LayoutSegment(CHAIN_REPLICATION, 0L, 21L,
+                Arrays.asList(stripe1, stripe2, stripe3));
+
+        LayoutStripe stripe11 = new LayoutStripe(Arrays.asList("C", "D"));
+        LayoutStripe stripe22 = new LayoutStripe(Collections.singletonList("C"));
+        LayoutStripe stripe33 = new LayoutStripe(Collections.singletonList("D"));
+        LayoutSegment segment2 = new LayoutSegment(CHAIN_REPLICATION, 21L, 51L,
+                Arrays.asList(stripe11, stripe22, stripe33));
+
+        List<LayoutSegment> layoutSegments = Arrays.asList(segment1, segment2);
+
+        Layout testLayout = createTestLayout(layoutSegments);
+
+        RedundancyCalculator calculator = new RedundancyCalculator("localhost");
+
+        ImmutableList<TransferSegment> transferSegments =
+                calculator.createStateList(testLayout, -1L);
+
+        MockedSegment presentSegment = new MockedSegment(0L,
+                20L,
+                TransferSegmentStatus.builder().segmentState(RESTORED).totalTransferred(21L).build());
+
+
+        MockedSegment nonPresentSegment = new MockedSegment(21L,
+                50L, TransferSegmentStatus.builder().segmentState(NOT_TRANSFERRED).totalTransferred(0L).build());
+
+        Assert.assertTrue(transformListToMock(transferSegments).containsAll(Arrays.asList(presentSegment,
+                nonPresentSegment)));
+
+        TransferSegmentStatus presentSegmentStatus = presentSegment.status;
+        assertThat(SegmentState.RESTORED)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unnecessary use of fully qualified name 'SegmentState.RESTORED' due to existing static import 'org.corfudb.infrastructure.log.statetransfer.StateTransferManager.TransferSegmentStatus.SegmentState.RESTORED'](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651453)